packages/contracts/contracts/deploy/PayableDAppFactory.sol

```diff
@@ -0,0 +1,155 @@
+// SPDX-License-Identifier: Apache-2.0
+/// @title ERC20 DApp Factory
+pragma solidity ^0.8.13;
+
+import {IConsensus} from "@cartesi/rollups/contracts/consensus/IConsensus.sol";
+import {CartesiDApp} from "@cartesi/rollups/contracts/dapp/CartesiDApp.sol";
+import {ICartesiDApp} from "@cartesi/rollups/contracts/dapp/ICartesiDApp.sol";
+import {ICartesiDAppFactory} from "@cartesi/rollups/contracts/dapp/ICartesiDAppFactory.sol";
+
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {Pausable} from "@openzeppelin/contracts/security/Pausable.sol";
+
+import {IFinancialProtocol} from "../protocol/IFinancialProtocol.sol";
+import {IMachineProtocol} from "../protocol/IMachineProtocol.sol";
+import {IPayableDAppFactory} from "./IPayableDAppFactory.sol";
+
+/// @notice This contract implements an IPayableDAppFactory based on a IERC20 token
+contract PayableDAppFactory is
+    IPayableDAppFactory,
+    IFinancialProtocol,
+    IMachineProtocol,
+    Ownable,
+    Pausable
+{
+    ICartesiDAppFactory public immutable factory;
+    IERC20 public immutable token;
+    IConsensus public immutable consensus;
+    uint256 public immutable price;
+
+    mapping(ICartesiDApp => uint256) public runway;
+
+    /// @notice Construct a new ERC20DAppFactory with the given token as payment method
+    /// @param _factory The factory to delegate calls to instantiate the dapp
+    /// @param _consensus The consensus contract to use
+    /// @param _price The price per second of execution of the dapp node
+    constructor(
+        address _owner,
+        ICartesiDAppFactory _factory,
+        IERC20 _token,
+        IConsensus _consensus,
+        uint256 _price
+    ) {
+        factory = _factory;
+        token = _token;
+        consensus = _consensus;
+        price = _price;
+
+        // transfer ownership to the give owner
+        transferOwnership(_owner);
+    }
+
+    /// @dev This function deploys a new application, and registers it with the payment system
+    function newApplication(
```

Comment:
Do you plan to add an entry point for deterministic deployment as well?

Reply:
I don't see a need now